packages/ui/src/Tooltip.tsx

```diff
@@ -1,37 +1,122 @@
+ * A tooltip with this content is displayed when user hovers over a target element.
+ *
+ * ### General info
+ * - Tooltip automatically detects viewport overflow and changes placement to prevent it.
+ * - Offset (space between a target and tooltip elements) can be also configured via "offset" property.
+ * - If "content" property is undefined, tooltip element will be removed from DOM entirely.
+ * - It's required to set the "id" property which will be assigned to the tooltip. This id parameter can be then used as a value of "aria-labelledby" attribute.
+ *
+ * ### Usage
+ * Wrap the target element with Tooltip component and use the "content" property to define what should be displayed inside the tooltip.
+ */
+export const Tooltip: FC<TooltipProps> = ({
+  content,
+  hideTimeoutDuration = 100,
+  offset = 10,
+  placement = 'top',
+  visible: visibilityOverride,
   children,
   ...props
 }) => {
-  return overlay ? (
-    <RCTooltip
-      overlay={wrapMaxWidth ? <div className={styles.wrapper}>{overlay}</div> : overlay}
-      overlayClassName={styles.overlay}
-      {...props}
-    >
-      {wrapElement ? (
-        // TODO: remove once https://github.com/react-component/tooltip/issues/18#issuecomment-411476678 has been resolved
-        <div style={{ display: 'inline-block' }}>{children}</div>
-      ) : (
-        children
+  const [isShown, setShown] = useState<boolean>(false)
+  const [hideTimeout, setHideTimeout] = useState<NodeJS.Timeout | null>(null)
+
+  /*
+   * We're using useState instead of useRef because react-popper package has some issues with useRef:
+   * https://github.com/popperjs/react-popper/issues/241#issuecomment-591411605
+   */
+  const [referenceElement, setReferenceElement] = useState<HTMLElement | null>(null)
+  const [popperElement, setPopperElement] = useState<HTMLSpanElement | null>(null)
+  const [arrowElement, setArrowElement] = useState<HTMLSpanElement | null>(null)
+
+  const { styles: popperStyles, attributes: popperAttributes } = usePopper(referenceElement, popperElement, {
+    placement,
+    modifiers: [
+      {
+        name: 'arrow',
+        options: {
+          element: arrowElement,
+        },
+      },
+      {
+        name: 'offset',
+        options: {
+          offset: [0, offset],
+        },
+      },
+    ],
+  })
+
+  const onMouseEnter = useCallback(() => {
+    if (hideTimeout !== null) {
+      clearTimeout(hideTimeout)
+    }
+
+    setShown(true)
+  }, [hideTimeout])
+
+  const onMouseLeave = useCallback(() => {
+    setHideTimeout(setTimeout(() => setShown(false), hideTimeoutDuration))
+  }, [hideTimeoutDuration])
+
+  useEffect(() => {
+    referenceElement?.addEventListener('mouseenter', onMouseEnter)
+    referenceElement?.addEventListener('mouseleave', onMouseLeave)
+
+    popperElement?.addEventListener('mouseenter', onMouseEnter)
+    popperElement?.addEventListener('mouseleave', onMouseLeave)
+  }, [referenceElement, popperElement, onMouseEnter, onMouseLeave])
```

Comment:
```suggestion
  useEffect(() => {
    referenceElement?.addEventListener('mouseenter', onMouseEnter)
    referenceElement?.addEventListener('mouseleave', onMouseLeave)

    popperElement?.addEventListener('mouseenter', onMouseEnter)
    popperElement?.addEventListener('mouseleave', onMouseLeave)
    
    return () => {
      referenceElement?.removeEventListener('mouseenter', onMouseEnter)
      referenceElement?.removeEventListener('mouseleave', onMouseLeave)

      popperElement?.removeEventListener('mouseenter', onMouseEnter)
      popperElement?.removeEventListener('mouseleave', onMouseLeave)
    }
  }, [referenceElement, popperElement, onMouseEnter, onMouseLeave])
```

Reply:
I would suggest to go a little further and split it into 2 `useEffect` statements: one for `onMouseLeave`, and another one for `onMouseEnter`. Otherwise, when `onMouseLeave` changes the subscription for `onMouseEnter` is also going to be updated.

Reply:
Thanks guys I totally missed this!

Reply:
Updated 👍